File: backend/domain/auth/operation.go
package auth

import (
	"context"

	"github.com/kanaru-ssk/payment-manager/backend/domain/emailaddress"
)

type Operation interface {
	// SendSignInLink: サインイン用リンク送信
	// email: メールアドレス
	SendSignInLink(ctx context.Context, email emailaddress.EmailAddress) error

	// SignInWithLink: リンクによるサインイン
	// email: メールアドレス
	// verificationToken: 検証トークン
	// return: 認証情報
	SignInWithLink(ctx context.Context, email emailaddress.EmailAddress, verificationToken string) (*Auth, error)
}

File: frontend/src/domain/auth/operation.ts
import type { Status } from "@/domain/status";
import type { SuccessOrError } from "@/lib/success-or-error";
import type { Auth } from "./entity";
import type { Email } from "./value-object";

/**
 * サインイン用リンク送信
 * @param email - メールアドレス
 */
export type SendSigninLink = (
	email: Email,
) => Promise<SuccessOrError<Status, Status>>;

/**
 * リンクによるサインイン
 * @param email - メールアドレス
 * @param verificationToken - 検証トークン
 * @returns - 認証情報
 */
export type SignInWithLink = (
	email: Email,
	verificationToken: string,
) => Promise<SuccessOrError<Auth, Status>>;
